src/Nethermind/Nethermind.Evm/VirtualMachine.cs

```diff
@@ -1246,12 +1246,73 @@ static void UpdateCurrentState(EvmState state, int pc, long gas, int stackHead)
                     }
                 case Instruction.EXTCODESIZE:
                     {
+                        Metrics.ExtCodeSizeOpcode++;
                         long gasCost = spec.GetExtCodeCost();
                         if (!UpdateGas(gasCost, ref gasAvailable)) goto OutOfGas;
 
                         Address address = stack.PopAddress();
                         if (!ChargeAccountAccessGas(ref gasAvailable, vmState, address, spec)) goto OutOfGas;
 
+                        if (programCounter < code.Length)
+                        {
+                            bool optimizeAccess = false;
+                            Instruction nextInstruction = (Instruction)code[programCounter];
+                            if (nextInstruction == Instruction.ISZERO)
+                            {
+                                optimizeAccess = true;
+                                Metrics.ExtCodeSizeOptimizedIsZero++;
+                            }
+                            else if ((nextInstruction == Instruction.GT || nextInstruction == Instruction.EQ) &&
+                                    stack.PeekUInt256IsZero())
```

Comment:
I missed this `Peek`. Maybe extract to a `var zeroOnTopOfStack` or something?

Reply:
Added comment